src/gt4py/next/program_processors/runners/dace_iterator/__init__.py

```diff
@@ -307,3 +312,177 @@ def build_sdfg_from_itir(
+            )
+        offset_provider = self.connectivities  # tables are None at this point
+
+        sdfg = self.backend.executor.otf_workflow.step.translation.generate_sdfg(  # type: ignore[union-attr]
+            self.itir,
+            arg_types,
+            offset_provider=offset_provider,
+            column_axis=kwargs.get("column_axis", None),
+        )
+
+        self.sdfg_closure_vars["sdfg.arrays"] = sdfg.arrays
+        if "storage" not in Program.connectivity_tables_data_descriptors:
+            for k in offset_provider.keys():
+                if not hasattr(offset_provider[k], "table"):
+                    continue
+                if connectivity_identifier(k) in sdfg.arrays:
+                    Program.connectivity_tables_data_descriptors["storage"] = sdfg.arrays[
+                        connectivity_identifier(k)
+                    ].storage
+                    break
+
+        # TODO(kotsaloscv): Keep halo exchange related metadata here?
+        # Could possibly be computed directly from the SDFG.
+
+        input_fields = [
+            str(inpt.id)
+            for closure in self.itir.closures
+            for inpt in closure.inputs
+            if str(inpt.id) in fields
+        ]
+        sdfg.gt4py_program_input_fields = {
+            inpt: dim
+            for inpt in input_fields
+            for dim in fields[inpt].dims  # type: ignore[union-attr]
+            if dim.kind == common.DimensionKind.HORIZONTAL
+        }
+
+        output_fields = []
+        for closure in self.itir.closures:
+            output = closure.output
+            if isinstance(output, itir.SymRef):
+                if str(output.id) in fields:
+                    output_fields.append(str(output.id))
+            else:
+                for arg in output.args:
+                    if str(arg.id) in fields:  # type: ignore[attr-defined]
+                        output_fields.append(str(arg.id))  # type: ignore[attr-defined]
+        sdfg.gt4py_program_output_fields = {
+            output: dim
+            for output in output_fields
+            for dim in fields[output].dims  # type: ignore[union-attr]
+            if dim.kind == common.DimensionKind.HORIZONTAL
+        }
+
+        sdfg.offset_providers_per_input_field = {}
+        itir_tmp = itir_transforms.apply_common_transforms(
+            self.itir, offset_provider=offset_provider
+        )
+        for closure in itir_tmp.closures:  # type: ignore[union-attr]
+            shifts = itir_transforms.trace_shifts.TraceShifts.apply(closure)
+            for k, v in shifts.items():  # type: ignore[assignment]
+                if not isinstance(k, str):
+                    continue
+                if k not in sdfg.gt4py_program_input_fields:
+                    continue
+                sdfg.offset_providers_per_input_field.setdefault(k, []).extend(list(v))
+
+        return sdfg
```

Comment:
It's hard to understand all the stuff that's going on in this function. I think it should be split in small parts implemented as free private functions with meaningful names that help to understand the main idea.

Reply:
I find it quite concise as is (~70 lines), and for this reason I would prefer not to break it further. I have added some comments to clarify the different steps and did some reductions/reshuffling ;) .